icelake/src/io/task_writer.rs

```diff
@@ -147,3 +169,216 @@ impl UnpartitionedWriter {
         self.data_file_writer.close().await
     }
 }
+
+/// Partition task writer
+pub struct PartitionedWriter {
+    schema: ArrowSchemaRef,
+    operator: Operator,
+    table_location: String,
+    location_generator: Arc<DataFileLocationGenerator>,
+    partition_spec: PartitionSpec,
+    partition_type: Any,
+    table_config: TableConfigRef,
+    /// # TODO
+    ///
+    /// Support to config the data file writer.
+    groups: HashMap<OwnedRow, PartitionGroup>,
+}
+
+struct PartitionGroup {
+    pub partition_array: StructArray,
+    pub writer: DataFileWriter,
+}
+
+impl PartitionedWriter {
+    /// Create a new `PartitionedWriter`.
+    pub fn new(
+        schema: ArrowSchemaRef,
+        table_location: String,
+        location_generator: DataFileLocationGenerator,
+        partition_spec: PartitionSpec,
+        partition_type: Any,
+        operator: Operator,
+        table_config: TableConfigRef,
+    ) -> Self {
+        Self {
+            schema,
+            operator,
+            table_location,
+            location_generator: location_generator.into(),
+            partition_spec,
+            partition_type,
+            table_config,
+            groups: HashMap::new(),
+        }
+    }
+
+    /// This function do two things:
+    /// 1. Partition the batch by partition spec.
+    /// 2. Create the partition group if not exist.
+    ///
+    /// # TODO
+    /// Mix up two thing in one function may be not a good idea.
+    /// The reason do that is we need to use the partition info in step 1 when we create the
+    /// partition group. To avoid data copy, we do that in one function. We may need to refactor
+    /// this function in the future. Also the name of this function may not be a good choice.
+    async fn split_by_partition(
+        &mut self,
+        batch: &RecordBatch,
+    ) -> Result<HashMap<OwnedRow, RecordBatch>> {
+        let value_array = Arc::new(StructArray::from(
+            self.partition_spec
+                .fields
+                .iter()
+                .map(|field| {
+                    let array: ArrayRef = batch
+                        .column_by_name(&field.name)
```

Comment:
We don't need to search by name for each partition, this should happen in initialization of writer.

Reply:
I'm not quit understand. Seems we can't do in initialization. For every batch coming in, we need to extract its related column to compute according partition field. And it's not gurantee that batch comes in is always have the same column order so we need to search it by name. (I'm not sure whether the function name is missleading.

Reply:
I think we should do it in initialization, and the column index should be found by schema. It's required that the record batch's schema should match table schema, otherwise the parquet file's schema doesn't match table schema.

Reply:
Oh, I see. Yes we should do it.